src/d3a/models/area/redis_dispatcher/market_event_dispatcher.py

```diff
@@ -0,0 +1,122 @@
+        # If there is no active trade but we still have deferred events, we need to handle these.
+        elif not self.active_trade and len(self.deferred_events) > 0:
+            self.store_deferred_events_during_active_trade(payload)
+            self.run_deferred_events()
+        # No deferred event, handle directly the event
+        else:
+            # Consume already deferred events before handling the new one
+            from concurrent.futures import ThreadPoolExecutor
+            executor = ThreadPoolExecutor(max_workers=1)
+
+            def executor_func():
+                self.root_dispatcher.event_listener(event_type=event_type, **kwargs)
+                self.publish_response(event_type)
+
+            executor.submit(executor_func)
+
+    def response_callback(self, payload):
+        data = json.loads(payload["data"])
+
+        if "response" in data:
+            event_type = data["response"]
+            event_type_id = data["event_type"]
+            if event_type not in self.str_market_events:
+                raise Exception("RedisAreaDispatcher: Should never reach this point")
+            else:
+                self.events_to_wait[MarketEvent(event_type_id)].set()
+
+    def publish_event(self, area_slug, event_type: MarketEvent, **kwargs):
+        dispatch_chanel = f"{area_slug}/market_event"
+
+        for key in ["offer", "trade", "new_offer", "existing_offer"]:
+            if key in kwargs:
+                kwargs[key] = kwargs[key].to_JSON_string()
+        send_data = {"event_type": event_type.value, "kwargs": kwargs}
+        self.redis.publish(dispatch_chanel, json.dumps(send_data))
+
+    @property
+    def _trade_related_events(self):
+        return [MarketEvent.OFFER_CHANGED, MarketEvent.TRADE,
+                MarketEvent.BID_TRADED, MarketEvent.BID_CHANGED]
+
+    def broadcast_event_redis(self, event_type: MarketEvent, **kwargs):
+        for child in sorted(self.area.children, key=lambda _: random()):
+            self.publish_event(child.slug, event_type, **kwargs)
+            self.events_to_wait[event_type].wait()
+            self.events_to_wait[event_type].clear()
+
+        for time_slot, agents in self.root_dispatcher._inter_area_agents.items():
+            if time_slot not in self.area._markets.markets:
+                # exclude past IAAs
+                continue
+
+            if not self.area.events.is_connected:
+                break
+            for area_name in sorted(agents, key=lambda _: random()):
+                agents[area_name].event_listener(event_type, **kwargs)
+
+    def publish_response(self, event_type):
+        response_channel = f"{self.area.parent.slug}/market_event_response"
+        response_data = json.dumps({"response": event_type.name.lower(),
+                                    "event_type": event_type.value})
+        self.redis.publish(response_channel, response_data)
+
+    def parse_market_event_from_event_payload(self, payload):
```

Comment:
This function is implemented twice (once here and once in MarketNotifyEventSubscriber. I suggest to move it outside and and import it. It also seams like a static method. Thanks

Reply:
Refactored it and moved it to market_structures.py